examples/feature_selection/plot_feature_selection_pipeline.py

```diff
@@ -1,40 +1,92 @@
 from sklearn.datasets import make_classification
-from sklearn.feature_selection import SelectKBest, f_classif
-from sklearn.pipeline import make_pipeline
 from sklearn.model_selection import train_test_split
-from sklearn.metrics import classification_report
 
-print(__doc__)
-
-# import some data to play with
 X, y = make_classification(
-    n_features=20, n_informative=3, n_redundant=0, n_classes=4,
+    n_features=20, n_informative=3, n_redundant=0, n_classes=2,
     n_clusters_per_class=2)
-
 X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
 
-# ANOVA SVM-C
-# 1) anova filter, take 3 best ranked features
-anova_filter = SelectKBest(f_classif, k=3)
-# 2) svm
-clf = svm.LinearSVC()
+# %%
+# A common mistake done with feature selection is indeed to search a subset
+# of discriminative features on the full dataset instead of only using the
+# training set. The usage of scikit-learn :func:`~sklearn.pipeline.Pipeline`
+# prevents to make such mistake.
+#
+# Here, we will demonstrate how to build a pipeline where the first step will
+# be the feature selection.
+
+# %%
+from sklearn.feature_selection import SelectKBest, f_classif
+from sklearn.pipeline import make_pipeline
+from sklearn.svm import LinearSVC
 
+anova_filter = SelectKBest(f_classif, k=3)
+clf = LinearSVC()
 anova_svm = make_pipeline(anova_filter, clf)
+anova_svm
+
+# %%
+# We created a pipeline where a feature selection will be the first step of
+# the processing before to use its output in an SVM classifier.
+# When calling `fit` on the training data, a subset of feature will be selected
+# and the index of these selected features will be stored. The feature selector
+# will subsequently reduce the number of feature and pass this subset to the
+# classifier which will be trained.
+
+# %%
 anova_svm.fit(X_train, y_train)
+
+# %%
+# Once the training accomplished, we can predict on new unseen samples. In this
+# case, the feature selector will only select the most discriminative features
+# based on the information stored during training. Then, the data will be
+# passed to the classifier which will make the prediction.
+#
+# Here, we report the final metrics via a classification report.
+
+# %%
+from sklearn.metrics import classification_report
+
 y_pred = anova_svm.predict(X_test)
 print(classification_report(y_test, y_pred))
 
-coef = anova_svm[:-1].inverse_transform(anova_svm['linearsvc'].coef_)
-print(coef)
+# %%
+# Be aware that you can inspect part of the pipeline. For instance, we might
+# be interested about the parameters of the classifier. Since we selected
+# three features, we expect to have three coefficients.
+
+# %%
+anova_svm[-1].coef_
+
+# %%
+# However, we do not know which features where selected from the original
+# dataset. We could proceed by several manner. Here, we will inverse the
+# transformation of these coefficients to get information about the original
+# space.
+
+# %%
+anova_svm[:-1].inverse_transform(anova_svm[-1].coef_)
```

Comment:
Side note: I am a little sad that we have to do this. I wonder why `_get_support_mask` is not public.

Reply:
feature names fix this also, right?

Reply:
This example was there originally. But to be honest I would have inspected the feature selector instead of back-propagating in this way.

I think that the features names will even make a better example.